pkg/util/flock/flock_unix.go

```diff
@@ -0,0 +1,51 @@
+// +build linux darwin freebsd openbsd netbsd dragonfly
+
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flock
+
+import (
+	"os"
+	"sync"
+
+	"golang.org/x/sys/unix"
+)
+
+var (
+	// lock guards lockfile. Assignment is not atomic.
+	lock sync.Mutex
+	// os.File has a runtime.Finalizer so the fd will be closed if the struct
+	// is garbage collected. Let's hold onto a reference so that doesn't happen.
+	lockfile *os.File
+)
+
+// Acquire acquires a lock on a file for the duration of the process. This method
+// is reentrant.
+func Acquire(path string) error {
+	lock.Lock()
+	defer lock.Unlock()
+	var err error
+	if lockfile, err = os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0666); err != nil {
```

Comment:
Thanks for explaining! Nice idea. I'm just a little skeptical about running
kubelet in a container as of now. This PR should help us identify the
issues with running kubelet in a container though.

On Mon, Feb 8, 2016 at 2:34 PM, Mike Danese <notifications@github.com>
wrote:

> In pkg/util/flock/flock_unix.go
> <https://github.com/kubernetes/kubernetes/pull/20253#discussion_r52236497>
> :
>
> > +
> > +var (
> > +	// lock guards lockfile. Assignment is not atomic.
> > +	lock sync.Mutex
> > +	// os.File has a runtime.Finalizer so the fd will be closed if the struct
> > +	// is garbage collected. Let's hold onto a reference so that doesn't happen.
> > +	lockfile *os.File
> > +)
> > +
> > +// Acquire acquires a lock on a file for the duration of the process. This method
> > +// is reentrant.
> > +func Acquire(path string) error {
> > +	lock.Lock()
> > +	defer lock.Unlock()
> > +	var err error
> > +	if lockfile, err = os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0666); err != nil {
>
> Yes probably. I will fix.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20253/files#r52236497>.
>